rfcs/79-minimal-fragmenting-relationships.md

```diff
@@ -0,0 +1,145 @@
+
+alice_mut.insert_pair::<Eats>(apple)
+	.insert_pair::<Eats>(banana);
+
+alice_mut.has_pair<Eats>(apple); // == true
+
+// Finds all (Eats, *) pairs and iterates the targets
+for target in alice_mut.targets<Eats>() {
+	// target == apple
+	// target == banana
+}
+
+// Matches all entities with at least one (Eats, *) pair and exposes the targets
+let query = world.query<(Entity, Targets<Eats>)>();
+for entity, targets in query.iter(&world) {
+	// entity == alice
+	for target in targets {
+		// target == apple
+		// target == banana
+	}
+}
+
+let query = world.query<Entity, With<Targets<Eats>>>();
+for entity in query.iter(&world) {
+	// entity == alice
+}
+
+```
+
+Dynamic queries would also be able to match on specific pairs:
+```rust
+let query = QueryBuilder::<Entity>::new(&mut World)
+	.with_pair::<Eats>(apple)
+	.build();
+
+query.single(&world); // == alice
+```
+
+As relationship pairs are treated internally as regular components this doesn't require any changes to the more complex query iteration APIs such as `par_iter`.
+## Implementation strategy
+
+There are a number of technical blockers before even a minimal version of relationships could be implemented in Bevy. They're individually broken down below:
+#### Archetype Cleanup
+In order to fit each pair into a single `Identifier` we lose the entity generation (see this [article](https://ajmmertens.medium.com/doing-a-lot-with-a-little-ecs-identifiers-25a72bd2647) for more details). As a consequence when an entity is despawned we need to ensure that all the archetypes that included that entity as a target are also destroyed so that a future entity re-using that id is not misinterpreted as being the old target. This is non-trivial.
+
+`bevy_ecs` currently operates under the assumption that archetype and component ids are dense and strictly increasing. In order to break that invariant we need to address several issues:
+##### Query and System Caches
+Query's caches of matching tables and archetypes are updated each time the query is accessed by iterating through the archetypes that have been created since it was last accessed. As the number of archetypes in bevy is expected to stabilize over the runtime of a program this isn't currently an issue. However with the increased archetype fragmentation caused by fragmenting relationships and the new wrinkle of archetype deletion the updating of the query caches is poised to become a performance bottleneck.
+
+An appropriate solution would allow us to expose new/deleted archetypes to only the queries that may be interested in iterating them. Luckily observers [#10839](https://github.com/bevyengine/bevy/pull/10839) could cleanly model this pattern. By sending ECS triggers each time an archetype is created and creating an observer for each query we can target the updates so that we no longer need to iterate every new archetype. Similarly we can fire archetype deletion events to cause queries to remove the archetype from their cache.
```

Comment:
No objections to this: supporting archetype cleanup is a valuable optimization no matter what we do.